src/orquestra/sdk/_base/_driver/_ce_runtime.py

```diff
@@ -399,28 +399,40 @@ def stop_workflow_run(
                 "- the authorization token was rejected by the remote cluster."
             ) from e
 
-    def list_workflow_run_summaries(
-        self,
-        *,
-        limit: Optional[int] = None,
-        max_age: Optional[timedelta] = None,
-        state: Optional[Union[State, List[State]]] = None,
-        workspace: Optional[WorkspaceId] = None,
-    ) -> List[WorkflowRunSummary]:
+    @staticmethod
+    def _list_wf_runs(
+        func,
+        limit: Optional[int],
+        max_age: Optional[timedelta],
+        state: Optional[Union[State, List[State]]],
+        workspace: Optional[WorkspaceId],
+    ) -> list:
         """
-        List summaries of the workflow runs, with some filters
+        Iterate through pages listing workflow runs.
+
+        This method encompasses the logic the drives `list_workflow_runs` and
+        `list_workflow_run_summaries`.
 
         Args:
-            limit: Restrict the number of runs to return, prioritising the most recent.
-            max_age: Only return runs younger than the specified maximum age.
-            status: Only return runs of runs with the specified status.
-            workspace: Only return runs from the specified workspace.
+            func: the paginated listing function over which to iterate. Must have a
+                signiture matching:
+
+                .. code-block::
+                    func(
+                        workflow_def_id: Optional[WorkflowDefID],
+                        page_size: Optional[int],
+                        page_token: Optional[str],
+                        workspace: Optional[WorkspaceId],
+                    )
+
+                and return a list.
```

Comment:
**non-blocking,suggestion**: have you considered using [Callable](https://docs.python.org/3/library/typing.html#annotating-callables)? Same benefits as using PEP484 type hints instead of verbal description.

Reply:
Protocols might be even a better tool in this situation, given that we call `func` via kwargs, and just the argument types don't convey enough information to be useful.

```
from typing import Protocol

class Paginator(Protocol):
    def __call__(
        self, 
        workflow_def_id: Optional[WorkflowDefID],
        page_size: Optional[int],
        page_token: Optional[str],
        workspace: Optional[WorkspaceId],
    )

...

    @staticmethod
    def _list_wf_runs(
        func: Paginator,
        ...
    )
```

Reply:
I've not used protocol before, that is a _nice_ facility to have! :D [f84dc1d](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/313/commits/f84dc1d55c2243b9bf16595a7c7933e18bbb5e64)